trie/triedb/pathdb/nodebufferlist.go

```diff
@@ -110,31 +115,220 @@ func newNodeBufferList(
 
-	go nf.loop()
+	var (
+		wg           sync.WaitGroup
+		start        = time.Now()
+		startStateID = nbl.persistID + 1
+	)
+	startBlock, err := readBlockNumber(freezer, startStateID)
+	if err != nil {
+		log.Error("Failed to read start block number", "error", err, "state_id", startStateID)
+		return nil, err
+	}
+	endBlock, err := readBlockNumber(freezer, freezerLength)
+	if err != nil {
+		log.Error("Failed to read end block number", "error", err, "state_id", freezerLength)
+		return nil, err
+	}
+	blockIntervals := nbl.createBlockInterval(startBlock, endBlock)
+	stateIntervals, err := nbl.createStateInterval(freezer, startStateID, freezerLength, blockIntervals)
+	if err != nil {
+		return nil, err
+	}
+	log.Info("block intervals info", "blockIntervals", blockIntervals, "stateIntervals", stateIntervals, "startBlock",
+		startBlock, "endBlock", endBlock)
+
+	nbl.linkMultiDiffLayers(len(blockIntervals))
+	errChan := make(chan error, 1)
+	for current, i := nbl.head, 0; current != nil; current, i = current.next, i+1 {
+		wg.Add(1)
+		go func(index int, slice []uint64, mdl *multiDifflayer) {
+			defer wg.Done()
+			for j := slice[0]; j <= slice[1]; j++ {
+				h, err := nbl.readStateHistory(freezer, j)
+				if err != nil {
+					errChan <- err
+				}
+				if h == nil {
+					return
+				}
+				if err = mdl.commit(h.meta.root, j, h.meta.block, 1, flattenTrieNodes(h.nodes)); err != nil {
+					log.Error("Failed to commit trie nodes to multi diff layer", "error", err)
+					errChan <- err
+				}
+			}
+		}(i, stateIntervals[i], current)
+	}
+	wg.Wait()
+	close(errChan)
+	if err = <-errChan; err != nil {
+		log.Error("Returned error from goroutine", "error", err)
+		return nil, err
+	}
 
-	log.Info("new node buffer list", "proposed block interval", nf.wpBlocks,
-		"reserve multi difflayers", nf.rsevMdNum, "difflayers in multidifflayer", nf.dlInMd,
-		"limit", common.StorageSize(limit), "layers", layers, "persist id", nf.persistID, "base_size", size)
-	return nf
+	for current, i := nbl.head, 0; current != nil; current, i = current.next, i+1 {
+		nbl.size += current.size
+		nbl.layers += current.layers
+	}
+	nbl.diffToBase()
+
+	log.Info("Succeed to add diff layer", "elapsed", common.PrettyDuration(time.Since(start)),
+		"base_size", nbl.base.size, "tail_state_id", nbl.tail.id, "head_state_id", nbl.head.id,
+		"nbl_layers", nbl.layers, "base_layers", nbl.base.layers)
+	return nbl, nil
+}
+
+// linkMultiDiffLayers links specified amount of multiDiffLayers for recovering
+func (nf *nodebufferlist) linkMultiDiffLayers(blockIntervalLength int) {
+	for i := 0; i < blockIntervalLength; i++ {
+		mdl := newMultiDifflayer(nf.limit, 0, common.Hash{}, make(map[common.Hash]map[string]*trienode.Node), 0)
+		nf.pushFront(mdl)
+	}
+	nf.count = uint64(blockIntervalLength)
+}
+
+func (nf *nodebufferlist) readStateHistory(freezer *rawdb.ResettableFreezer, stateID uint64) (*history, error) {
+	h, err := readHistory(freezer, stateID)
+	if err != nil {
+		if checkError(err) {
```

Comment:
`checkError` func is confused. if return true, it's not error, return false is a real error. 
It's better to refine it.

Reply:
This function is used to check wether there are state data in ancient db.